web3/utils/subscriptions.py

```diff
@@ -0,0 +1,285 @@
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Coroutine,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Sequence,
+    TypeVar,
+    Union,
+)
+
+from eth_typing import (
+    Address,
+    ChecksumAddress,
+    HexStr,
+)
+from hexbytes import (
+    HexBytes,
+)
+
+from web3.exceptions import (
+    Web3AttributeError,
+    Web3ValueError,
+)
+from web3.types import (
+    BlockData,
+    FilterParams,
+    LogReceipt,
+    SyncProgress,
+    TxData,
+)
+
+if TYPE_CHECKING:
+    from web3 import (
+        AsyncWeb3,
+    )
+    from web3.providers.persistent.subscription_manager import (
+        SubscriptionManager,
+    )
+    from web3.types import EthSubscriptionResult  # noqa: F401
+
+
+TSubscriptionResult = TypeVar("TSubscriptionResult", bound="EthSubscriptionResult")
+TSubscription = TypeVar("TSubscription", bound="EthSubscription[Any]")
+
+
+class EthSubscriptionContext(Generic[TSubscription, TSubscriptionResult]):
+    def __init__(
+        self,
+        async_w3: "AsyncWeb3",
+        subscription: TSubscription,
+        result: TSubscriptionResult,
+        **kwargs: Any,
+    ) -> None:
+        self.async_w3 = async_w3
+        self.subscription = subscription
+        self.result = result
+        self.__dict__.update(kwargs)
+
+    def __getattr__(self, item: str) -> Any:
+        if item in self.__dict__:
+            return self.__dict__[item]
+        raise Web3AttributeError(
+            f"'{self.__class__.__name__}' object has no attribute '{item}'"
+        )
+
+
+EthSubscriptionHandler = Callable[
+    [EthSubscriptionContext[Any, Any]], Coroutine[Any, Any, None]
+]
+
+
+def handler_wrapper(
+    handler: Optional[EthSubscriptionHandler],
+) -> Optional[EthSubscriptionHandler]:
+    """Wrap the handler to add bookkeeping and context creation."""
+    if handler is None:
+        return None
+
+    async def wrapped_handler(
+        context: EthSubscriptionContext[TSubscription, TSubscriptionResult],
+    ) -> None:
+        sx = context.subscription
+        sx.handler_call_count += 1
+        sx.manager.total_handler_calls += 1
```

Comment:
It's very possible I missed this, but other than testing, where does this get used?

Reply:
It's simply for keeping track. As a user I'd like to know these stats from a subscription manager if it's doing some background work that I can't see. If you think we should remove it, I can do so.